### PR TITLE
[Improvements]: Bitcoin OutPoint

### PR DESCRIPTION
**Pull request by Milerius**
## Description

- noexcept
- add equality_comparable.h / equality_comparable_tests.cpp for CRTP relational injections
- use operator == is fine for std::array because internally it use std::equal and that take cares of the size too.
- use RAII constructor style by integrating to_array constexpr template function
- make the copy constructor call the constructor number 1 to avoid code repetition and enjoy RAII
- use struct instead of class if all members are public

## How to test

```bash
make -C build tests && build/tests/tests tests
```

## Types of changes

Modernize code (use modern c++, improve readability, more safeness)

## Checklist

- [x] Create pull request as draft initially, unless its complete.
- [x] Add/Update tests to cover changes as needed.
- [x] Update documentation as needed.
- [x] If there is a related Issue, mention it in the description.

ps: depends on #2317

### CODE REVIEW

**Comment by Milerius** [comment]
@catenocrypt @hewigovens When I run with coverage all the UT I see that the relational operators are never used: 

Capture d’écran 2022-06-15 à 17 10 33

I'm also able to compile without:
Capture d’écran 2022-06-15 à 17 12 47

Should we delete them or it's a false positive ?

**Comment by hewigovens** [comment]
probably not used at all

**Comment by hewigovens** [comment]
@Milerius ready to merge?

File: src/algorithm/to_array.h [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+}
+
+template <std::size_t N, typename U = void, typename Iter,
+          typename V = typename std::iterator_traits<Iter>::value_type,
+          typename T = std::conditional_t<std::is_same<U, void>{}, V, U>>
+constexpr auto to_array(Iter iter) -> std::array<T, N> {
+    return to_array<T>(iter, std::make_index_sequence<N>{});
+}
+
+} // namespace TW
```
Comment (hewigovens) [inline src/algorithm/to_array.h:27]: can you add a new line for new added files? others look good

**Review by hewigovens — APPROVED** [review]

File: src/algorithm/to_array.h (hunk 1 above)
Comment (Milerius) [reply src/algorithm/to_array.h]: https://github.com/trustwallet/wallet-core/pull/2319/commits/01b811b250a65c6cf766a52ec1f8f47f025de35b

**Comment by hewigovens** [comment]
@Milerius could you please check CI on master? we have clang tidy/asm enabled, it failed after this pr: https://github.com/trustwallet/wallet-core/runs/7009222189?check_suite_focus=true

**Comment by Milerius** [comment]
> @Milerius could you please check CI on master? we have clang tidy/asm enabled, it failed after this pr: https://github.com/trustwallet/wallet-core/runs/7009222189?check_suite_focus=true

Checking now

**Comment by Milerius** [comment]
> @Milerius could you please check CI on master? we have clang tidy/asm enabled, it failed after this pr: https://github.com/trustwallet/wallet-core/runs/7009222189?check_suite_focus=true

Fixed in #2340 - didn't knew we had a configuration with ASAN, I run all the unit tests now with it enabled too now.

**Comment by hewigovens** [comment]
Yeah, because it's very slow to run, so we didn't enable it for normal pr

**Pull request merged**